Rework pkg/loopback to fix #39155 and mitigate loop module race conditions

djsweet opened the pull request:
The root cause of issue #39155 is that `pkg/loopback` attempts to open
a loop device, the `loop` module allocates a new loop device because all
others are busy, but the container is not seeing an appropriate
`/dev/loopX` device being created. Inside the `moby` unit test
container, there is no `udev`, so the natural consequence is that the
`loop` module communicates a free device index that is not visible to
the tests.

This reworking of `pkg/loopback` fixes #39155, and handles some other outstanding issues as well:

1. The `/dev` filesystem is no longer probed scanning for open loop
   devices; this reworking appropriately uses the `loop` module
   userspace API to determine the next free index. (Previously, this
   API was not being used correctly, hence the need for `/dev` scanning)
2. `loop` device nodes are forcibly created if the `loop` module reports
   them as free but they do not exist on the filesystem, solving a race
   condition between the daemon and `udev` even outside of unit testing.
3. Race conditions between determining the next free device index and
   acquiring said device for file attachment are handled by way of
   random sleep intervals, which makes the race condition mitigation
   appropriate for both in-process races and environmental races, e.g.
   some other OCI implementation uses the `loop` module at the exact
   same time.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Internally reworked `pkg/loopback` to fix #39155, without altering the external API.

**- How I did it**
- Fixed `ioctlLoopCtlGetFree` to use the `loop` module `ioctl` interface properly, so that the next free loop device index is actually captured.
- Created new file `loop_module.go` with an interface and implementation to automatically create a device node for the reported next free loop device index if a suitable node does not already exist.
- Replaced file scan loop in `openNextAvailableLoopback` with a retry system based on random sleep intervals, so that races between next-free and open are mitigated by non-overlapping sleep intervals.

**- How to verify it**
Unit tests for the new interfaces have been provided for both device node creation and race condition mitigation; and are automatically run as part of the unit test suite.

**- Description for the changelog**
Reworked `pkg/loopback` to automatically create device nodes, eliminating `udev` race conditions; also mitigating race conditions around attaching to loop devices.

djsweet committed:
Use correct `LOOP_CTL_GET_FREE` API in `pkg/loopback`

The `ioctl` interface for the `LOOP_CTL_GET_FREE` request on
`/dev/loop-control` is a little different from what `unix.IoctlGetInt`
expects: the first index is the returned status in `r1`, not an `int`
pointer as the first parameter.

Unfortunately we have to go a little lower level to get the appropriate
loop device index out, using `unix.Syscall` directly to read from
`r1`. Internally, the index is returned as a signed integer to match the
internal `ioctl` expectations of interpreting a negative signed integer
as an error at the userspace ABI boundary, so the direct interface of
`ioctlLoopCtlGetFree` can remain as-is.

Signed-off-by: Daniel Sweet <danieljsweet@icloud.com>

djsweet committed:
Ham-fist create `/dev/loop` nodes in `pkg/loopback`

We are now at a place where we can reliably determine which loop device
is the "next" free one, but in some situations the "next" free loop
device isn't visible in our `/dev` namespace by the time we want to use
it. There are at least two underlying reasons for this:

1. `udev` isn't running and our `/dev` namespace is static, for some
   underlying reason (e.g. testing inside a Docker container)
2. `udev` has lost the race to create the device node, but we just so
   happen to need it immediately.

By forcibly creating the device node ourselves, we cleanly handle both
cases. The only downside for the second case is that `udev` may report
an error if the device node already exists.

Signed-off-by: Daniel Sweet <danieljsweet@icloud.com>

djsweet committed:
Refactor `loop` module operations into testable interface/impl design

There's a lot that can go wrong with the simple act of trying to
establish a `loop` device attachment, so it's for the better that the
internals of the package are easily testable with mock implementations.

Signed-off-by: Daniel Sweet <danieljsweet@icloud.com>

djsweet committed:
Replace per-file retry loop with random sleep attempt loop in `loopback`

The previous behavior of `loopback` in the event that it could not open
an appropriate loop device relied on the "next free device" being 0 and
looping through all numeric entries of `/dev/loopX` until either an open
device was found or no devices were available. Because the internals of
`loopback` now report the "next free device" appropriately, this no
longer functions as expected: only the devices above the "next free device"
were going to be visited, which in the usual case of one remaining free
device means that failures would occur more often.

There are two levels of race condition we have to handle around
acquiring the next open device now: process-level and environment-level.
Locking would eliminate process-level race conditions, but
environment-level race conditions would not be handled. Therefore, the
`loopback` package handles all race conditions as if they were
environment-level, which incidentally also mitigates process-level race
conditions.

These race conditions are mitigated from the in-process perspective by
using 32 retries and a random sleep interval bounded between 0ms and 64ms.
In the absolute worst-case, the `loopback` module will now block for
~2 seconds, but only if more than 32 concurrent attempts to open the
same "free" device occur, which is guaranteed to succeed with five 9s
of reliability for 33 attempts due to the choice of sleep interval and
retry periods. This also functions properly for one-off environment
conditions, where some other process is attaching files to loop devices
concurrently.

Signed-off-by: Daniel Sweet <danieljsweet@icloud.com>

djsweet committed:
Add test for `loopback` AttachToNextAvailableDevice

The functionality behind OpenSysfsParameterFile,
GetMaxPartitionParameter, GetMknodDeviceNumber, MakeIndexNode, and
AttachToNextAvailableDevice is not stubbed out. The race tests cover
PerformPathStat, GetNextFreeDeviceIndex, GetDeviceBaseNodeStat,
PerformMknod, and OpenDeviceFile depending on whether enough loop
devices are already available. So, this just about covers it for the
loop attach alteration.

Signed-off-by: Daniel Sweet <danieljsweet@icloud.com>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "fix-39155-with-loopback-improvements" git@github.com:djsweet/moby.git somewhere
$ cd somewhere
$ git commit --amend -s --no-edit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

thaJeztah commented:
ping @kolyshkin PTAL; looks like this one is up your alley

djsweet commented:
I've been working on figuring out exactly _why_ so many builds failed on Jenkins, and I've come to discover something rather unfortunate, after reverting my patch back to 0c07dad66a09b393165b5f6c44d8f7eda2a722cb:

I have 100% success rate on `make test` when attached directly to a TTY and 100% failure rate on `make test` when attached to a pipe. E.g., `make test | cat` fails 100% of the time on 0c07dad66a09b393165b5f6c44d8f7eda2a722cb for me. The failures I'm getting are pretty disparate, here's what I've seen so far:

1. `TestServiceRemoveKeepsIngressNetwork`:
`service_test.go:265: timeout hit after 30s: waiting for all tasks to be removed: task count at 1`
(I'm pretty sure this is tracked by #37306)
2. Various `test timed out after 5m0s` failures across the CLI test suites, it's never the same test
3. Various `Log in groutine after X has completed` messages in the daemon test suites, and like the timeouts it's never the same test

Again, this is after rolling back to 0c07dad66a09b393165b5f6c44d8f7eda2a722cb so I think I might have to make another issue entirely for those build failures.

kolyshkin commented:
> Created new file loop_module.go with an interface and implementation to automatically create a device node for the reported next free loop device index if a suitable node does not already exist.

@djsweet I took a quick look, it seems that either loop_module.go contains more than needed to solve the task you're describing above, or there are too many public functions, or perhaps I need to read it more carefully. Some of these are used only for testing, so maybe move them to `attach_loopback_test.go` for the sake of easier review.

kolyshkin commented:
Also, if it's possible to split this into a few commits, each one addressing one particular problem, that is also welcomed.

djsweet committed:
Make `golint` happy with `pkg/loopback`

The biggest change here is that `attachErrorStateString` is moved to
`attach_loopback_test.go`, as it's not used for anything but test
debugging.

Signed-off-by: Daniel Sweet <danieljsweet@icloud.com>

djsweet committed:
Clean up `loopModuleContext` interface and associated tests

Now that the tests are existent and cover most of the functionality of
opening a new loop device, much of the interface surface area of
loopModuleContext proved to be unnecessary. Much of the former interface
methods now have their concrete implementations used directly by the
remaining interface implementations in `concreteLoopModuleContext`.

As an additional change, all methods of `loopModuleContext` are now
lower-cased. While the interface itself was never exported, lower-casing
all of the implementation methods makes the non-exported nature of the
interface more readily apparent.

Signed-off-by: Daniel Sweet <danieljsweet@icloud.com>

djsweet committed:
Clean up formatting in `pkg/loopback`

Formatting in `pkg/loopback` is once again compliant with what is
expected by `go fmt`.

Signed-off-by: Daniel Sweet <danieljsweet@icloud.com>

djsweet committed:
Actually use `loopControl` constant, remove redundant typecasts, break

All of these issues weren't picked up by the default `golint`, but were
picked up on the build system.

Signed-off-by: Daniel Sweet <danieljsweet@icloud.com>

djsweet commented:
@kolyshkin The first iteration of this PR was first a series of commits, then squash merged into a branch based off of `master`. I've re-pushed with the original series of commits; it might offer some better insight as to why the internals are the way they are.

You definitely had a point about there being excess structure: the `loopModuleContext` interface only really exists to facilitate mocking out non-idempotent IO code, so I've now cleaned up the interface to only expose the operations that are mocked in the unit tests. The interface methods are also now non-exported, in addition to the interface itself being non-exported, to clarify in their usage that they're meant for internal consumption only.

This latest `experimental` build is still suffering from the `make test | cat` failure problem. I'm still working on finding out which commit caused this to start happening. Not sure about `janky` and `z` this time around; I'm currently running integration tests locally to see if I can't reproduce those failures.

thaJeztah commented:
ping @kolyshkin PTAL

kolyshkin commented on pkg/loopback/attach_loopback.go:
```diff
@@ -38,13 +42,126 @@ func getNextFreeLoopbackIndex() (int, error) {
+
+	maxPart, err := strconv.ParseUint(scanner.Text(), 10, 0)
+	return uint(maxPart), err
+}
```
I suggest to use `ioutil.ReadFile()` here, otherwise it's too complicated.

kolyshkin replied:
```go
	buf, err := ioutil.ReadFile("/sys/module/loop/parameters/max_part")
	if err != nil {
		// This parameter is expected to exist for the foreseeable future
		// but it wouldn't hurt to handle the case where it's missing.
		if os.IsNotExist(err) {
			return 0, nil
		}
		return 0, err
	}
	maxPart, err := strconv.ParseUint(buf, 10, 0)
	return uint(maxPart), err
}

kolyshkin commented on pkg/loopback/attach_loopback.go:
```diff
@@ -38,13 +42,126 @@ func getNextFreeLoopbackIndex() (int, error) {
+
+	minor := index << partShift
+	return ((minor & minorHighMask) << minorHighShift) | loopMajorMask | (minor & minorLowMask), nil
+}
```
Please use `unix.Mkdev()` instead, no need to reinvent wheels :)

kolyshkin commented on pkg/loopback/attach_loopback.go:
```diff
@@ -38,13 +42,126 @@ func getNextFreeLoopbackIndex() (int, error) {
+
+	if err = syscall.Mknod(loopPath, uint32(baseStats.Mode|syscall.S_IFBLK), deviceNumber); err != nil {
+		// If the mknod call failed because it already exists, we're fine
+		if asErrno, ok := err.(syscall.Errno); !ok || asErrno != syscall.EEXIST {
```
Use `if os.IfExist(err)` here.

kolyshkin commented on pkg/loopback/attach_loopback.go:
```diff
@@ -38,13 +42,126 @@ func getNextFreeLoopbackIndex() (int, error) {
+		return nil, err
+	}
+
+	if err = syscall.Mknod(loopPath, uint32(baseStats.Mode|syscall.S_IFBLK), deviceNumber); err != nil {
```
might change to `unix.Mknod` since we're using `unix` package already

kolyshkin commented on pkg/loopback/loop_module.go:
```diff
@@ -0,0 +1,177 @@
+
+func (ctx *concreteLoopModuleContext) MakeIndexNode(index int) (os.FileInfo, error) {
+	return directIndexMknod(ctx, index)
+}
```
nit: missing `\n` at EOF

kolyshkin commented on pkg/loopback/loop_module.go:
```diff
@@ -0,0 +1,177 @@
+	PerformMknod(path string, mode uint32, dev int) error
+	MakeIndexNode(index int) (os.FileInfo, error)
+}
+
```
Are you introducing this interface in order to be able to test the package? Problem is, interfaces are slow, so I would not use it unless we're going to abstract various loopback types (we don't).

kolyshkin commented:
@djsweet thanks for your work! Can you please make a separate PR out of the first commit, this one is absolutely ready to be merged? We can work on the rest of it here.

kolyshkin commented:
I am going to close this one in favor of https://github.com/moby/moby/pull/39801 and https://github.com/moby/moby/pull/40056.

Reasons are:
 - it's too much changes for the obsoleted graphdriver; 
 - we expect all systems to have working udev (so no manual device creation is needed);
 - for unit test, the fix belong to the test code (see https://github.com/moby/moby/pull/40056).
